src/libopensc/card-piv.c

```diff
@@ -467,82 +791,1819 @@ static struct sc_card_driver piv_drv = {
+		}
+	}
+	return SC_SUCCESS;
+}
+#endif
+
+static int piv_load_options(sc_card_t *card)
+{
+	int r;
+	piv_private_data_t * priv = PIV_DATA(card);
+	size_t i, j;
+	scconf_block **found_blocks, *block;
+
+	const char *option = NULL;
+	int piv_max_object_size_found = 0;
+#ifdef ENABLE_PIV_SM
+	int piv_pairing_code_found = 0;
+	int piv_use_sm_found = 0;
+#endif
+
+	if ((option = getenv("PIV_MAX_OBJECT_SIZE")) != NULL) {
+		sc_log(card->ctx, "getenv(\"PIV_MAX_OBJECT_SIZE\")=\"%s\"", option);
+		priv->max_object_size = atoi(option);
+		if (priv->max_object_size < PIV_MAX_OBJECT_SIZE || priv->max_object_size > MAX_FILE_SIZE) {
+			sc_log(card->ctx,"Invalid max_object_size: \"%d\"", priv->max_object_size);
+			if (priv->max_object_size < PIV_MAX_OBJECT_SIZE)
+				priv->max_object_size  = PIV_MAX_OBJECT_SIZE;
+			else
+				priv->max_object_size = MAX_FILE_SIZE; /* conserative value if error */
+		} else
+			piv_max_object_size_found = 1;
+		sc_log(card->ctx," priv->max_object_size:%d", priv->max_object_size);
+	}
+
+#ifdef ENABLE_PIV_SM
+	/* pairing code is 8 decimal digits and is card specific */
+	if ((option = getenv("PIV_PAIRING_CODE")) != NULL) {
+		sc_log(card->ctx,"getenv(\"PIV_PAIRING_CODE\") found");
+		if (piv_parse_pairing_code(card, option) == SC_SUCCESS) {
+			memcpy(priv->pairing_code, option, PIV_PAIRING_CODE_LEN);
+			piv_pairing_code_found = 1;
+		}
+	}
+
+	if ((option = getenv("PIV_USE_SM"))!= NULL) {
+		sc_log(card->ctx,"getenv(\"PIV_USE_SM\")=\"%s\"", option);
+		if (!strcmp(option, "never")) {
+			priv->sm_flags |= PIV_SM_FLAGS_NEVER;
+			piv_use_sm_found = 1;
+		}
+		else if (!strcmp(option, "always")) {
+			priv->sm_flags |= PIV_SM_FLAGS_ALWAYS;
+			piv_use_sm_found = 1;
+		}
+		else {
+			sc_log(card->ctx,"Invalid piv_use_sm: \"%s\"", option);
+		}
+	}
+#endif
+
+	for (i = 0; card->ctx->conf_blocks[i]; i++) {
+		found_blocks = scconf_find_blocks(card->ctx->conf, card->ctx->conf_blocks[i],
+				"card_driver", "PIV-II");
+		if (!found_blocks)
+			continue;
+
+		for (j = 0, block = found_blocks[j]; block; j++, block = found_blocks[j]) {
+
+#ifdef ENABLE_PIV_SM
+			/*
+			 * "piv_use_sm" if card supports NIST sp800-73-4 sm, when should it be used
+			 * never - use card like 800-73-3, i.e. contactless is very limited on
+			 * true PIV cards. Some  PIV-like" card may allow this.
+			 * this security risk
+			 * always - Use even for contact interface.
+			 * PINS, crypto and reading of object will not show up in logs
+			 * or over network.
+			 */
+
+			if (piv_use_sm_found == 0) {
+				option = scconf_get_str(block, "piv_use_sm", "default");
+				sc_log(card->ctx,"conf: \"piv_use_sm\"=\"%s\"", option);
+				if  (strcmp(option,"default")) {
+					if (!strcmp(option, "never")) {
+						priv->sm_flags |= PIV_SM_FLAGS_NEVER;
+					}
+					else if (!strcmp(option, "always")) {
+						priv->sm_flags |= PIV_SM_FLAGS_ALWAYS;
+					}
+					else {
+						sc_log(card->ctx,"Invalid piv_use_sm: \"%s\"", option);
+					}
+				}
```

Comment:
Is there a specific reason why this condition is constructed this way and not as a linear if/else if/else other than the "default" is no-op? I would argue that it would be more readable if it would not be nested.

Reply:
fixed. I think I have fixed all but two, where I commented why I did not change see 9c68797bb6b732c7a1de883584e73ffb052d8552

Reply:
Fixed in 796141eee